src/vm/builtins/builtin_runner/poseidon.zig

```diff
@@ -54,37 +69,362 @@ pub const PoseidonBuiltinRunner = struct {
+    ///
+    /// # Parameters
+    ///
+    /// - `segments`: A pointer to the `MemorySegmentManager` for segment management.
+    /// - `pointer`: A `Relocatable` pointer to the current stack pointer.
+    ///
+    /// # Returns
+    ///
+    /// A `Relocatable` pointer to the final stack pointer, or an error code if the
+    /// verification fails.
+    pub fn finalStack(
+        self: *Self,
+        segments: *MemorySegmentManager,
+        pointer: Relocatable,
+    ) !Relocatable {
+        if (self.included) {
+            const stop_pointer_addr = pointer.subUint(1) catch return RunnerError.NoStopPointer;
+            const stop_pointer = try (segments.memory.get(stop_pointer_addr) orelse return RunnerError.NoStopPointer).tryIntoRelocatable();
+
+            if (@as(
+                i64,
+                @intCast(self.base),
+            ) != stop_pointer.segment_index) {
+                return RunnerError.InvalidStopPointerIndex;
+            }
+
+            const stop_ptr = stop_pointer.offset;
+            if (stop_ptr != try self.getUsedInstances(segments) * @as(
+                usize,
+                @intCast(self.cells_per_instance),
+            )) {
+                return RunnerError.InvalidStopPointer;
+            }
+            self.stop_ptr = stop_ptr;
+            return stop_pointer_addr;
+        }
+
+        self.stop_ptr = 0;
+        return pointer;
+    }
+
+    /// Get the number of used cells associated with this Keccak runner.
+    ///
+    /// # Parameters
+    ///
+    /// - `segments`: A pointer to a `MemorySegmentManager` for segment size information.
+    ///
+    /// # Returns
+    ///
+    /// The number of used cells as a `u32`, or `MemoryError.MissingSegmentUsedSizes` if
+    /// the size is not available.
+    pub fn getUsedCells(self: *const Self, segments: *MemorySegmentManager) !u32 {
+        return segments.getSegmentUsedSize(
+            @intCast(self.base),
+        ) orelse MemoryError.MissingSegmentUsedSizes;
+    }
+
+    pub fn getUsedInstances(self: *Self, segments: *MemorySegmentManager) !usize {
+        return std.math.divCeil(
+            usize,
+            try self.getUsedCells(segments),
+            @intCast(self.cells_per_instance),
+        );
+    }
+
     pub fn deduceMemoryCell(
-        self: *const Self,
+        self: *Self,
+        allocator: Allocator,
         address: Relocatable,
         memory: *Memory,
-    ) ?MaybeRelocatable {
-        _ = memory;
-        _ = address;
-        _ = self;
-        return null;
+    ) !?MaybeRelocatable {
+        const index = @mod(
+            @as(
+                usize,
+                @intCast(address.offset),
+            ),
+            @as(
+                usize,
+                @intCast(self.cells_per_instance),
+            ),
+        );
+        if (index < @as(
+            usize,
+            @intCast(self.n_input_cells),
```

Comment:
Here as `n_input_cells` is u32 and you want to cast to usize, I don't think that you have to write that, type coercion should be automatic and implicit with

```zig
if (index < self.n_input_cells)
```